From 0ef11365c086d3e93a46a3c8316c727447c1c775 Mon Sep 17 00:00:00 2001
From: Alona Solntseva <alona.solntseva@tandemg.com>
Date: Sun, 29 May 2022 09:28:39 +0300
Subject: [PATCH] [SW-0] tpm: fix pcr extend failure in erom

Change sha type from 384 to 256 to match tpm pcr algo

Signed-off-by: Alona Solntseva <alona.solntseva@tandemg.com>
Change-Id: Ic2cb7397dbfdd5f7a6a3543fb75feda80f651bbf
---

diff --git a/app/ext_rom/Makefile b/app/ext_rom/Makefile
index e064f69..7b1a31a 100644
--- a/app/ext_rom/Makefile
+++ b/app/ext_rom/Makefile
@@ -105,7 +105,7 @@
 	$(PREBOOT_ROOT_DIR)/lib/common/spi_flash.o \
 	$(PREBOOT_ROOT_DIR)/lib/common/boot_if.o \
 	$(PREBOOT_ROOT_DIR)/lib/common/pll.o \
-	$(PREBOOT_ROOT_DIR)/lib/common/sha512.o \
+	$(PREBOOT_ROOT_DIR)/lib/common/sha256.o \
 	$(PREBOOT_ROOT_DIR)/lib/common/efuse.o \
 	$(PREBOOT_ROOT_DIR)/lib/common/lib_common.o \
 	$(HL_DEV_ARCH_COMMON)/tpm2.o \
diff --git a/app/ext_rom/erom_tpm.c b/app/ext_rom/erom_tpm.c
index 2afd4fc..bb1e4b7 100644
--- a/app/ext_rom/erom_tpm.c
+++ b/app/ext_rom/erom_tpm.c
@@ -110,8 +110,8 @@
 	/* offset of ppboot+pcie_fw image to verify, start from fw_version */
 	ppboot_pcie_offset = (HL_FWSRAM_BASE + IMG_LEN_SIZE);
 
-	*offset = ppboot_pcie_size;
-	*size = ppboot_pcie_offset;
+	*offset = ppboot_pcie_offset;
+	*size = ppboot_pcie_size;
 
 	return 0;
 }
diff --git a/arch/gaudi2/arc/common.c b/arch/gaudi2/arc/common.c
index 2e982df..5b89433 100644
--- a/arch/gaudi2/arc/common.c
+++ b/arch/gaudi2/arc/common.c
@@ -32,7 +32,7 @@
 #include <arcboot_export.h>
 #include <preboot.h>
 #include <pll.h>
-#include <sha.h>
+#include <sha256.h>
 #include <uart.h>
 #include <efuse.h>
 #include <string.h>
@@ -478,7 +478,7 @@
 	u32 flash_offset;
 	u32 num_cycles;
 	int ret;
-	uint8_t hash[SHA384_SUM_LEN];
+	uint8_t hash[SHA256_SUM_LEN];
 
 	if (image_ptr == ARCBOOT_PRIMARY) {
 		if (!validate_secondary_ppboot_fw())
@@ -548,7 +548,7 @@
 	ret = tpm_init();
 	if (!ret) {
 		memcpy(hash, (void *)(SDRAM_BASE + IMG_LEN_SIZE + image_size),
-		       SHA384_SUM_LEN);
+		       SHA256_SUM_LEN);
 		if (tpm_pcr_extend(TPM2_PREBOOT_PCR_IDX, hash))
 			preboot_puts("PPBOOT: pcr extend failed\n");
 	} else if (ret == -EOPNOTSUPP) {
diff --git a/arch/gaudi2/common/tpm_auth.c b/arch/gaudi2/common/tpm_auth.c
index d3e801e..81eab9e 100644
--- a/arch/gaudi2/common/tpm_auth.c
+++ b/arch/gaudi2/common/tpm_auth.c
@@ -9,7 +9,7 @@
 #include <boot_if.h>
 #include <tpm2.h>
 #include <uart.h>
-#include <sha.h>
+#include <sha256.h>
 #include <boot_desc.h>
 #include <errno.h>
 #include "lzo/unaligned.h"
@@ -95,7 +95,7 @@
 	if (hash == NULL)
 		return -1;
 
-	ret = tpm2_pcr_extend(image_index, TPM2_ALG_SHA256, hash, TPM2_SHA384_DIGEST_SIZE);
+	ret = tpm2_pcr_extend(image_index, TPM2_ALG_SHA256, hash, SHA256_SUM_LEN);
 	if (ret) {
 		preboot_puts("preboot_pcr_extend: pcr extend failed\n");
 		cpu_boot_err_set(CPU_BOOT_ERR_TPM_FAIL);
@@ -106,11 +106,18 @@
 
 int tpm_hash_and_pcr_extend(uint32_t image_index, uintptr_t image_ptr, uint32_t image_size)
 {
-	uint8_t hash[SHA_SUM_LEN] = { 0 };
+	int ret;
+	uint8_t hash[SHA256_SUM_LEN] = { 0 };
 
-	sha_csum((void *)image_ptr, image_size, hash);
+	sha256_csum((void *)image_ptr, image_size, hash);
 
-	return tpm_pcr_extend(image_index, hash);
+	ret = tpm2_pcr_extend(image_index, TPM2_ALG_SHA256, hash, SHA256_SUM_LEN);
+	if (ret) {
+		preboot_puts("preboot_pcr_extend: pcr extend failed\n");
+		cpu_boot_err_set(CPU_BOOT_ERR_TPM_FAIL);
+	}
+
+	return ret;
 }
 
 void preboot_tpm_auth(void)
